cirq/neutral_atoms/neutral_atom_devices.py

```
@@ -0,0 +1,424 @@
+        hasMeasurement = False
+        controlled_qubits_lists = []
+
+        # Count the number of each type of gate occuring in this moment.
+        # Also verify that gates of the same type are equal
+        for op in operations:
+            if isinstance(op, (ops.GateOperation,
+                               ops.ParallelGateOperation)):
+                if isinstance(op.gate, ops.ZPowGate):
+                    if num_parallel_z == 0:
+                        refZgate = op.gate
+                    else:
+                        if refZgate != op.gate:
+                            raise ValueError("Non-identical Parallel Z gates")
+                    num_parallel_z += len(op.qubits)
+
+                if isinstance(op.gate, (ops.XPowGate,
+                                        ops.YPowGate,
+                                        ops.PhasedXPowGate)):
+                    if num_parallel_xy == 0:
+                        refXYgate = op.gate
+                    else:
+                        if refXYgate != op.gate:
+                            raise ValueError("Non-identical Parallel XY gates")
+                    num_parallel_xy += len(op.qubits)
+
+                if isinstance(op.gate, (ops.CNotPowGate,
+                                        ops.CZPowGate,
+                                        ops.CCXPowGate,
+                                        ops.CCZPowGate)):
+                    if num_parallel_controlled_qubits == 0:
+                        refCgate = op.gate
+                    else:
+                        if refCgate != op.gate:
+                            raise ValueError("Non-identical Parallel Controlled"
+                                             " Gates")
+                    num_parallel_controlled_qubits += len(op.qubits)
+                    controlled_qubits_lists.append(op.qubits)
+                if isinstance(op.gate, ops.MeasurementGate):
+                    hasMeasurement = True
```

Should the validation fail if an unknown gate or operation is seen?

---

The super call iterates over every option to make sure that the device recognizes each operation.